src/cascadia/TerminalCore/TerminalApi.cpp

```diff
@@ -298,10 +298,57 @@ void Terminal::UseMainScreenBuffer()
     CATCH_LOG();
 }
 
+// NOTE: This is the version of AddMark that comes from VT
 void Terminal::AddMark(const Microsoft::Console::VirtualTerminal::DispatchTypes::ScrollMark& mark)
 {
     const til::point cursorPos{ _activeBuffer().GetCursor().GetPosition() };
     AddMark(mark, cursorPos, cursorPos);
+    auto* last = &_scrollMarks.back();
+    if (last->category == Microsoft::Console::VirtualTerminal::DispatchTypes::MarkCategory::Prompt)
+    {
+        _currentPrompt = last;
+    }
+}
+
+void Terminal::CommandStart()
+{
+    if (_currentPrompt)
+    {
+        const til::point cursorPos{ _activeBuffer().GetCursor().GetPosition() };
+
+        // Move the end of this mark to the current cursor position. The prompt
+        // is between [mark.start, mark.end)
+        _currentPrompt->end = cursorPos;
+    }
+}
+
+void Terminal::OutputStart()
+{
+    if (_currentPrompt)
+    {
+        const til::point cursorPos{ _activeBuffer().GetCursor().GetPosition() };
+        // Mark the current cursor pos as the the end of the command. The command
+        // is between [mark.end, mark.commandEnd)
+        _currentPrompt->commandEnd = cursorPos;
+    }
+}
+
+void Terminal::CommandFinished(std::optional<unsigned int> error)
+{
+    if (_currentPrompt)
+    {
+        const til::point cursorPos{ _activeBuffer().GetCursor().GetPosition() };
+        // Mark the current cursor pos as the the end of the output. The command
+        // is between [mark.commandEnd, mark.outputEnd)
+        _currentPrompt->outputEnd = cursorPos;
+
+        if (error.has_value())
+        {
+            _currentPrompt->category = *error == 0u ?
+                                           DispatchTypes::MarkCategory::Success :
+                                           DispatchTypes::MarkCategory::Error;
+        }
+    }
 }

```

Comment:
I think this code would be more robust if you implemented it as a small state machine or similar and maybe replace the 3 functions with a single "state transition" function. Because then we could be more confident that the code is robust against invalid VT sequences and that it maintains a proper state at all times.

Also, it would lend itself to get rid of the unsafe `_currentPrompt` pointer. That one is somewhat dangerous, since any rogue `_scrollMarks` modification can make turn it into a dangling pointer. Unless I'm mistaken, the above `Terminal::AddMark` modification is proof of that: It calls the other `AddMark` overload which calls `_currentPrompt.push_back()`, which might reallocate the vector's backing buffer. Any existing pointer into the vector would then be a dangling pointer, and since `_currentPrompt = last` is only done conditionally, this would potentially make `_currentPrompt` point into deallocated (or since reused) memory. The above `_currentPrompt->` writes would then overwrite random memory. (I'm blocking this PR over this issue.)

Because of that I'd suggest changing this into a state machine and get rid of `_currentPrompt`. Instead I'd only call `AddMark` once the state machine returns into the "ground" state.

Reply:
That's a couple good points. 

* [ ] The first being the dangling pointer. That should definitely be fixed
* [ ] The invalid state thing. I totally spaced, the way this is now, an 133;A followed by a 133;D creates a mark with the start=end, no commandEnd, and an outputEnd. I wonder what the behavior should be in that case? Should that all be treated as command? as output? as prompt? Should the `outputEnd` just get ignored? (probably that last one)
* [ ] The state machine thing I might follow up with you on teams about. Not sure I have enough brain cells currently to picture what you had in mind here.